cbits/dns.c

```diff
@@ -3,74 +3,71 @@
+// error code. See: https://msdn.microsoft.com/en-us/library/windows/desktop/ms681382(v=vs.85).aspx
+DWORD getWindowsDefDnsServers(char* dnsAddresses, size_t bufferLen) {
     FIXED_INFO *pFixedInfo;
     ULONG ulOutBufLen;
     DWORD dwRetVal;
 
-    dns_t* result = (dns_t*) MALLOC(sizeof(dns_t));
-    if (result == NULL) return (dns_t*) ERROR_NOT_ENOUGH_MEMORY;
+    if (bufferLen <= 0) return ERROR_NOT_ENOUGH_MEMORY;
 
-    result->dnsAddresses = (char*) MALLOC(128 * sizeof(char));
-    if (result->dnsAddresses == NULL) {
-      result->error = ERROR_NOT_ENOUGH_MEMORY;
-      return result;
-    }
-
-    result->error = NO_ERROR;
     pFixedInfo = (FIXED_INFO *) MALLOC(sizeof (FIXED_INFO));
-    if (pFixedInfo == NULL) {
-      result->error = ERROR_NOT_ENOUGH_MEMORY;
-      return result;
-    }
+    if (pFixedInfo == NULL)
+      return ERROR_NOT_ENOUGH_MEMORY;
     ulOutBufLen = sizeof (FIXED_INFO);
 
     // Make an initial call to GetAdaptersInfo to get the necessary size into the
     // ulOutBufLen variable
     if (GetNetworkParams(pFixedInfo, &ulOutBufLen) == ERROR_BUFFER_OVERFLOW) {
         FREE(pFixedInfo);
         pFixedInfo = (FIXED_INFO *) MALLOC(ulOutBufLen);
-        if (pFixedInfo == NULL) {
-            result->error = ERROR_NOT_ENOUGH_MEMORY;
-            return result;
-        }
+        if (pFixedInfo == NULL)
+            return ERROR_NOT_ENOUGH_MEMORY;
     }
 
     dwRetVal = GetNetworkParams(pFixedInfo, &ulOutBufLen);
 
     if (dwRetVal == NO_ERROR) {
       int offset = 0;
-      int space_available = 255;
+      int spaceAvailable = bufferLen;
       IP_ADDR_STRING* head = &pFixedInfo->DnsServerList;
-      while (head != NULL && space_available >= 16) {
-        int ip_len = strlen(head->IpAddress.String);
-        strcpy_s(result->dnsAddresses + offset, ip_len + 1, head->IpAddress.String);
-        // Update the offset (non-null-terminated IP address + separator)
-        offset += ip_len + 1;
-        // Update the space available, pessimistically
-        space_available -= 16;
+
+      while (head != NULL && spaceAvailable >= 16) {
+        int ipLen = strlen(head->IpAddress.String);
+        int copySize = ipLen + 1;
+
+        // Copy the IP address, including the null terminator.
+        strcpy_s(dnsAddresses + offset, copySize, head->IpAddress.String);
+
+        spaceAvailable -= copySize;
+        if (spaceAvailable <= 0) break;
+
+        offset += copySize;
+
         // Write the separator, but only if this is not the last one,
-        // otherwise terminate the string.
+        // otherwise the string is already terminated due to the call to
+        // strcpy_s, which copies the null terminator.
+
         head = head->Next;
-        if (head == NULL)
-          result->dnsAddresses[offset] = '\0';
-        else
-          result->dnsAddresses[offset - 1] = ',';
+        if (head != NULL)
+          dnsAddresses[offset - 1] = ',';
       }
+
     }
 
     else {
```

Comment:
I think this `else` is not necessary if we return `dwRetVal` at line 70 since `dwRetVal` is `NO_ERROR` in the normal case.